src/constraints/simplex.rs

```diff
@@ -0,0 +1,80 @@
+use super::Constraint;
+
+#[derive(Copy, Clone)]
+/// A simplex with alpha 'a', that is,
+/// a set of the form {x: sum(x) = a, x>=0} where alpha is a positive constant.
+pub struct Simplex {
+    alpha: f64,
+}
+
+impl Simplex {
+    /// Construct a new simplex with given (positive) alpha
+    pub fn new(alpha: f64) -> Self {
+        assert!(alpha > 0.0);
+        Simplex { alpha }
+    }
+}
+
+impl Constraint for Simplex {
+    fn project(&self, x: &mut [f64]) {
+        let a = &self.alpha;
+
+        // ---- step 1
+        let mut v = vec![x[0]]; // vector containing x[0]
+        let mut v_size_old: i64 = -1; // 64 bit signed int
+        let mut v_tilde: Vec<f64> = Vec::new(); // empty vector of f64
+        let mut rho: f64 = x[0] - a; // 64 bit float
+
+        // ---- step 2
+        x.iter().skip(1).for_each(|x_n| {
+            if *x_n > rho {
+                rho += (*x_n - rho) / ((v.len() + 1) as f64);
+                if rho > *x_n - a {
+                    v.push(*x_n);
+                } else {
+                    v_tilde.extend(&v);
```

Comment:
How can you pre-allocate an array if you don't know how much memory to pre-allocate it (I remember you saying it could be done but don't know how) - I would test with this and correct the issue later ?